pkg/sessionctx/variable/session.go

```diff
@@ -2428,8 +2429,22 @@ func (s *SessionVars) SetEnablePseudoForOutdatedStats(val bool) {
 	s.EnablePseudoForOutdatedStats = val
 }
 
-// GetReplicaRead get ReplicaRead from sql hints and SessionVars.replicaRead.
+// GetReplicaRead get ReplicaRead from sql hints and SessionVars.replicaRead with adjusted.
 func (s *SessionVars) GetReplicaRead() kv.ReplicaReadType {
+	// For test purpose, you can enable this failpoint to get the unadjusted replica read.
+	failpoint.Inject("GetReplicaReadUnadjusted", func(_ failpoint.Value) {
+		failpoint.Return(s.replicaRead)
+	})
+	// Replica read only works for read-only statements.
+	if !s.StmtCtx.IsReadOnly {
+		if s.StmtCtx.HasReplicaReadHint {
+			s.StmtCtx.AppendWarning(errors.New("Ignore replica read hint for non-read-only statement"))
```

Comment:
Seems `GetReplicaRead` has some side effects to append a warning. Could we estimate it only to append warning if necessary?

Reply:
It appends warning when user specify a hint which will be ignored.

`select /*+ READ_CONSISTENT_REPLICA() */ * from t where id = 1 for update;`

There is a problem that this function can be call multi times in a query, we don't want to send repeated warnings.

Reply:
Check the exist warnings before appending new one. The implementation is not such good but I don’t want to introduce too many changes for this usecase which is not a good practice. PTAL again.